File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys

# 상하좌우 이동 방향
Direction = [(-1, 0), (1, 0), (0, -1), (0, 1)]

# 단지 내 집 개수를 구하는 함수
def CountHouses(x, y):
    count = 1
    mapList[x] = mapList[x][:y] + '0' + mapList[x][y+1:]    # 해당 위치 방문 처리
    
    for dirX, dirY in Direction:
        newX = x + dirX
        newY = y + dirY

        # 새 좌표가 범위 안에 있고 해당 좌표에 집이 있다면
        if newX in range(n) and newY in range(n) and mapList[newX][newY] == '1':
            count += CountHouses(newX, newY)
    
    return count

n = int(sys.stdin.readline())
mapList = [ sys.stdin.readline().rstrip() for _ in range(n) ]

complex = 0   # 총 단지 수
count = []  # 단지별 가구 수
for x in range(n):
    for y in range(n):
        # 해당 위치에 집이 있다면
        if mapList[x][y] == '1':
            complex += 1  # 단지 수 1 증가
            count.append(CountHouses(x, y))

count.sort()
print(complex)
for n in count: print(n)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
const readline = require("readline");

let n;
const map = [];

const countHome = (startX, startY) => {
    let count = 1;
    const queue = [[startX, startY]];

    const direction = [[0, 1], [0, -1], [1, 0], [-1, 0]];
    let i = 0;
    while (i < queue.length) {
        const [x, y] = queue[i++];

        for (const [dx, dy] of direction) {
            const newX = x + dx;
            const newY = y + dy;

            if (newX < 0 || newX >= n || newY < 0 || newY >= n) {
                continue;
            }
            if (!map[newX][newY]) {
                continue;
            }

            map[newX][newY] = 0;
            count++;
            queue.push([newX, newY]);
        }
    }

    return count;
};

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});
rl.on("line", (line) => {
    if (!n) {
        n = Number(line);
    } else {
        map.push(line.split("").map(Number));
        if (map.length === n) {
            rl.close();
        }
    }
}).on("close", () => {
    let total = 0;
    const homeList = [];

    for (let x = 0; x < n; x++) {
        for (let y = 0; y < n; y++) {
            if (map[x][y]) {
                total++;
                map[x][y] = 0;
                homeList.push(countHome(x, y));
            }
        }
    }

    console.log(total);
    console.log(homeList.sort((a, b) => a - b).join("\n"));

    process.exit();
});